terminal-emulator/src/main/java/com/termux/terminal/TerminalEmulator.java

```diff
@@ -0,0 +1,2447 @@
+                // position on the preceding line. If the active position is at the top margin, a scroll down is performed".
+                if (mCursorRow <= mTopMargin) {
+                    mScreen.blockCopy(0, mTopMargin, mColumns, mBottomMargin - (mTopMargin + 1), 0, mTopMargin + 1);
+                    blockClear(0, mTopMargin, mColumns);
+                } else {
+                    mCursorRow--;
+                }
+                break;
+            case 'N': // SS2, ignore.
+            case '0': // SS3, ignore.
+                break;
+            case 'P': // Device control string
+                mOSCOrDeviceControlArgs.setLength(0);
+                continueSequence(ESC_P);
+                break;
+            case '[':
+                continueSequence(ESC_CSI);
+                break;
+            case '=': // DECKPAM
+                setDecsetinternalBit(DECSET_BIT_APPLICATION_KEYPAD, true);
+                break;
+            case ']': // OSC
+                mOSCOrDeviceControlArgs.setLength(0);
+                continueSequence(ESC_OSC);
+                break;
+            case '>': // DECKPNM
+                setDecsetinternalBit(DECSET_BIT_APPLICATION_KEYPAD, false);
+                break;
+            default:
+                unknownSequence(b);
+                break;
+        }
+    }
+
+    /** DECSC save cursor - http://www.vt100.net/docs/vt510-rm/DECSC . See {@link #restoreCursor()}. */
+    private void saveCursor() {
+        SavedScreenState state = (mScreen == mMainBuffer) ? mSavedStateMain : mSavedStateAlt;
+        state.mSavedCursorRow = mCursorRow;
+        state.mSavedCursorCol = mCursorCol;
+        state.mSavedEffect = mEffect;
+        state.mSavedForeColor = mForeColor;
+        state.mSavedBackColor = mBackColor;
+        state.mSavedDecFlags = mCurrentDecSetFlags;
+        state.mUseLineDrawingG0 = mUseLineDrawingG0;
+        state.mUseLineDrawingG1 = mUseLineDrawingG1;
+        state.mUseLineDrawingUsesG0 = mUseLineDrawingUsesG0;
+    }
+
+    /** DECRS restore cursor - http://www.vt100.net/docs/vt510-rm/DECRC. See {@link #saveCursor()}. */
+    private void restoreCursor() {
+        SavedScreenState state = (mScreen == mMainBuffer) ? mSavedStateMain : mSavedStateAlt;
+        setCursorRowCol(state.mSavedCursorRow, state.mSavedCursorCol);
+        mEffect = state.mSavedEffect;
+        mForeColor = state.mSavedForeColor;
+        mBackColor = state.mSavedBackColor;
+        int mask = (DECSET_BIT_AUTOWRAP | DECSET_BIT_ORIGIN_MODE);
+        mCurrentDecSetFlags = (mCurrentDecSetFlags & ~mask) | (state.mSavedDecFlags & mask);
+        mUseLineDrawingG0 = state.mUseLineDrawingG0;
+        mUseLineDrawingG1 = state.mUseLineDrawingG1;
+        mUseLineDrawingUsesG0 = state.mUseLineDrawingUsesG0;
+    }
+
+    /** Following a CSI - Control Sequence Introducer, "\033[". {@link #ESC_CSI}. */
+    private void doCsi(int b) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 253 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.